### ISSUE TITLE
[Design] Improve address scanning policy

### ISSUE THREAD

**Issue by pedroferreira1** [HathorNetwork/hathor-wallet-headless#95]
# Motivation

Some use cases (e.g. exchange integrations) have a large set of users that might (or not) use the address associated to him. With that in mind, the gap limit is not the best scanning policy we could use to determine when to stop loading transactions from the full node because of two problems (i) we end up loading many more addresses we need for safety reasons, and (ii) even with a safe number of gap limit, the user count might reach the maximum number of addresses generated because none of the last X users have used their addresses, then the only way to increase the gap limit is to stop the wallet, change the configuration and start again.

# Proposed solution

The proposed solution is to create a new configuration parameter in the headless called `scanningPolicy`, that for now we have two options, (i) `'gap-limit'`, or (ii) `'address-limit'`.

The `gap-limit` policy is the default one and the `address-limit` is the new one we will introduce in this design. If the user configures to use the `address-limit` policy, it requires a new configuration `scanLastIndex`, which is the number of addresses to load by default when starting the wallet.

This configuration is for the whole wallet headless, then if you'd like to start many wallets in the same headless server, you will use the same scan policy for all of them.

The new scan policy will work with the following rules:

1. When the wallet starts it will load only until the address with index = `scanLastIndex`.
2. We won't have a limit for the user request `/wallet/address`, we will always return a new one and load it in the full node.
3. We will have a new API that will only work with this scan policy called `/wallet/load-addresses?lastIndex=200` which loads the addresses from the last loaded index until the index in the GET parameter `lastIndex`.
4. We will add a new API to return the last loaded index that will only work with the new scan policy

**Comment by msbrogli** [comment]
I agree with the problem and the high level solution, but I feel we shouldn't put `scanLastIndex` in the config file because it is supposed to change over time. I feel we should add parameter in the wallet initialization API called `scanLastIndex` and the user can always request more and more addresses. The current addresses API could always return the index and the derivation path so the user can store both in their local database.

I like the new API `/wallet/lastAddressIndex` and it seems useful even when the policy is `gap-limit`.

Maybe we should persist the latest state of the wallet when it is running and it should be loaded the next time the wallet is started.

**Comment by pedroferreira1** [comment]
> I agree with the problem and the high level solution, but I feel we shouldn't put scanLastIndex in the config file because it is supposed to change over time. I feel we should add parameter in the wallet initialization API called scanLastIndex and the user can always request more and more addresses.

I agree that this should be a parameter in the `/start` POST request, not in the configuration file.

> The current addresses API could always return the index and the derivation path so the user can store both in their local database.

You mean the `/addresses` or `/address` API? I think both should return this but for compatibility reasons I think I will add the index and path only in the `/address`, which is the most useful in my opinion, what do you think? 

> I like the new API /wallet/lastAddressIndex and it seems useful even when the policy is gap-limit.

Agreed.

> Maybe we should persist the latest state of the wallet when it is running and it should be loaded the next time the wallet is started.

What do you mean by 'state'? If it's the last loaded index, I agree and it's in the design. However my idea was not to automatically load from the latest state but to warn the user if the new start will load all the addresses from the last load.

**Comment by msbrogli** [comment]
In regards to the state, I think that we should always load it by default and just notify when the file has not been found and a new one will be created.

For, this high level design is approved but we still need a low level design to discuss the data persistence.

**Comment by andreabadesso** [comment]
The design looks good to me

In regards to loading the entire state (loaded addresses) vs only the latest index from the disk, I think I agree with @pedroferreira1 but we can discuss this in more depth on the low level design

**Comment by pedroferreira1** [comment]
This design is approved and will be implemented in the future
